Java async profiler do not generate profiles on kubernetes

Hey

I am using coroot-node-agent 1.34.2 with `--enable-java-async-profiler` flag enabled. During initial testing with docker compose on my mac M3 Pro, profiler generated proper java profiles for dummy tomcat app. 

When I upgraded coroot node agent version to 1.34.2  and added `--enable-java-async-profiler` in kubernetes dev environment no additional profiles from async profiles appear on java applications (Old CPU profiles collected by epbf was present)

I have exec into few java containers and I see libasyncProfiler.so` in their /tmp/coroot ` is present. There are no any other files in `/tmp/coroot/` directory. 

```
bash-5.1$ ls /tmp/coroot/
libasyncProfiler.so
```

In logs of java container I see following error

```
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active                                                                                                                                                                                                                                             
[ERROR] Profiler is not active
```


When I filter out output of `coroot-node-agent` container with `async` keyword

```
apps-us-west1-dev-01$ kubectl logs -n ebpf-profiler-dev ebpf-profiler-5lxws | grep async
I0729 13:35:32.954770 2865162 async_profiler.go:88] pid=23290: async-profiler started (itimer+alloc+lock)
I0729 13:35:32.954798 2865162 profiling.go:243] pid=23290: async-profiler started
I0729 13:35:32.957857 2865162 async_profiler.go:88] pid=2847701: async-profiler started (itimer+alloc+lock)
I0729 13:35:32.957879 2865162 profiling.go:243] pid=2847701: async-profiler started
I0729 13:35:32.960517 2865162 async_profiler.go:88] pid=2194737: async-profiler started (itimer+alloc+lock)
I0729 13:35:32.960535 2865162 profiling.go:243] pid=2194737: async-profiler started
I0729 13:35:32.962698 2865162 async_profiler.go:88] pid=1726884: async-profiler started (itimer+alloc+lock)
I0729 13:35:32.962713 2865162 profiling.go:243] pid=1726884: async-profiler started
I0729 13:35:32.966345 2865162 async_profiler.go:88] pid=3665903: async-profiler started (itimer+alloc+lock)
I0729 13:35:32.966360 2865162 profiling.go:243] pid=3665903: async-profiler started
I0729 13:35:32.968488 2865162 async_profiler.go:88] pid=3677021: async-profiler started (itimer+alloc+lock)
I0729 13:35:32.968504 2865162 profiling.go:243] pid=3677021: async-profiler started
I0729 13:35:33.001248 2865162 async_profiler.go:88] pid=60752: async-profiler started (itimer+alloc+lock)
I0729 13:35:33.001292 2865162 profiling.go:243] pid=60752: async-profiler started
I0729 13:35:33.003498 2865162 async_profiler.go:88] pid=2847592: async-profiler started (itimer+alloc+lock)
I0729 13:35:33.003531 2865162 profiling.go:243] pid=2847592: async-profiler started
I0729 13:35:33.006843 2865162 async_profiler.go:88] pid=3439031: async-profiler started (itimer+alloc+lock)
I0729 13:35:33.006862 2865162 profiling.go:243] pid=3439031: async-profiler started
I0729 13:35:33.008845 2865162 async_profiler.go:88] pid=1524413: async-profiler started (itimer+alloc+lock)
I0729 13:35:33.008864 2865162 profiling.go:243] pid=1524413: async-profiler started
I0729 13:35:33.010926 2865162 async_profiler.go:88] pid=22995: async-profiler started (itimer+alloc+lock)
I0729 13:35:33.010943 2865162 profiling.go:243] pid=22995: async-profiler started
I0729 13:35:33.056272 2865162 async_profiler.go:88] pid=2114918: async-profiler started (itimer+alloc+lock)
I0729 13:35:33.056301 2865162 profiling.go:243] pid=2114918: async-profiler started
```

When I filter out output of `coroot-node-agent` container with `jvm` keyword

```
apps-us-west1-dev-01$ kubectl logs -n ebpf-profiler-dev ebpf-profiler-5lxws | grep -i jvm
I0729 13:34:39.636166 2865162 profiling.go:454] JVM detected PID: 2114918, hotspot: true, perfmap: false
I0729 13:34:39.746885 2865162 profiling.go:454] JVM detected PID: 22995, hotspot: true, perfmap: false
I0729 13:34:43.933344 2865162 profiling.go:454] JVM detected PID: 3665903, hotspot: true, perfmap: false
I0729 13:34:46.528522 2865162 profiling.go:454] JVM detected PID: 60752, hotspot: true, perfmap: false
I0729 13:34:59.633618 2865162 profiling.go:454] JVM detected PID: 3677021, hotspot: true, perfmap: false
I0729 13:34:59.739595 2865162 profiling.go:454] JVM detected PID: 1726884, hotspot: true, perfmap: true
I0729 13:35:04.566104 2865162 profiling.go:454] JVM detected PID: 2847701, hotspot: true, perfmap: true
I0729 13:35:04.763742 2865162 profiling.go:454] JVM detected PID: 1524413, hotspot: true, perfmap: true
I0729 13:35:05.740262 2865162 profiling.go:454] JVM detected PID: 2847592, hotspot: true, perfmap: false
I0729 13:35:05.958975 2865162 profiling.go:454] JVM detected PID: 2194737, hotspot: true, perfmap: true
I0729 13:35:06.040852 2865162 profiling.go:454] JVM detected PID: 3439031, hotspot: true, perfmap: true
I0729 13:35:06.657539 2865162 profiling.go:454] JVM detected PID: 23290, hotspot: true, perfmap: false
```



I've added asprof binary from async-profiler to coroot-node-agent and manually executed 


```
[root@ebpf-profiler-fdbgr /tmp] gke-apps-us-west1-de-n2d64m256d200-v1-fb28391c-bc2v $ asprof -d 30 -f /tmp/flamegraph.html --libpath /tmp/coroot/libasyncProfiler.so 2552710
Profiling for 30 seconds
Done
[root@ebpf-profiler-fdbgr /tmp]
```

`flamegraph.html` was available `/proc/2552710/root/tmp/` and contain proper data.

Could you help me investigate why coroot-node-agent java profiling do not work on kubernetes? Does it requires any additional configuration from java apps or additional configuration for coroot-node-agent pod?

Currently `coroot-node-agent` has

```
hostPID: true
securityContext: {
         runAsUser: 0,
         privileged: true,
}

```

--- a/jvm/async_profiler.go
+++ b/jvm/async_profiler.go
@@ -64,8 +64,14 @@ func IsAsyncProfilerAlreadyLoaded(pid uint32) bool {
 }
 
 func DeployAndStartAsyncProfiler(pid uint32) error {
-	if err := os.MkdirAll(proc.Path(pid, "root", apDeployDir), 0755); err != nil {
+	deployDir := proc.Path(pid, "root", apDeployDir)
+	if err := os.MkdirAll(deployDir, 0755); err != nil {
 		return fmt.Errorf("failed to create deploy dir: %w", err)
+	}
+	// The JFR file is created by the JVM process itself, which may run as a
+	// non-root user, so the dir must be world-writable (sticky, like /tmp).
+	if err := os.Chmod(deployDir, 0777|os.ModeSticky); err != nil {
+		return fmt.Errorf("failed to chmod deploy dir: %w", err)
 	}
 	libData, err := agentAssets.ReadFile("assets/" + apAssetName())
 	if err != nil {
@@ -96,23 +102,24 @@ func CollectAsyncProfiler(pid uint32) ([]byte, error) {
 	}
 	hostPath := apHostDumpPath(pid, nsPid)
 
-	if err := apCommand(pid, "stop"); err != nil {
-		return nil, fmt.Errorf("asprof stop failed for JVM %d: %w", pid, err)
-	}
+	stopErr := apCommand(pid, "stop")
 
-	data, err := os.ReadFile(hostPath)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return nil, nil
-		}
-		return nil, fmt.Errorf("failed to read async-profiler dump %s: %w", hostPath, err)
-	}
+	data, readErr := os.ReadFile(hostPath)
 	os.Remove(hostPath)
 
 	if err := apCommand(pid, apStartArgs(nsPid)); err != nil {
 		klog.Warningf("pid=%d: failed to restart async-profiler: %v", pid, err)
 	}
 
+	if stopErr != nil {
+		return nil, fmt.Errorf("asprof stop failed for JVM %d: %w", pid, stopErr)
+	}
+	if readErr != nil {
+		if os.IsNotExist(readErr) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to read async-profiler dump %s: %w", hostPath, readErr)
+	}
 	return data, nil
 }
 

--- a/jvm/jattach.go
+++ b/jvm/jattach.go
@@ -7,6 +7,7 @@ import (
 	"io"
 	"net"
 	"os"
+	"strconv"
 	"strings"
 	"syscall"
 	"time"
@@ -82,10 +83,7 @@ func LoadAgent(pid uint32, agentPath, args string) error {
 	if err != nil {
 		return err
 	}
-	if status != '0' {
-		return fmt.Errorf("load agent failed: status=%c response=%s", status, resp)
-	}
-	return nil
+	return checkAgentLoadResponse(status, resp)
 }
 
 func LoadNativeAgent(pid uint32, agentPath, args string) error {
@@ -94,8 +92,21 @@ func LoadNativeAgent(pid uint32, agentPath, args string) error {
 	if err != nil {
 		return err
 	}
+	return checkAgentLoadResponse(status, resp)
+}
+
+func checkAgentLoadResponse(status byte, resp string) error {
 	if status != '0' {
-		return fmt.Errorf("load native agent failed: status=%c response=%s", status, resp)
+		return fmt.Errorf("agent load failed: status=%c response=%s", status, resp)
+	}
+	line, _, _ := strings.Cut(resp, "\n")
+	line = strings.TrimSpace(strings.TrimPrefix(line, "return code: "))
+	code, err := strconv.Atoi(line)
+	if err != nil {
+		return fmt.Errorf("agent load failed: unexpected response %q", resp)
+	}
+	if code != 0 {
+		return fmt.Errorf("agent load failed: Agent_OnAttach returned %d", code)
 	}
 	return nil
 }

